Federated Model (seq-to-seq)/train_model.py

```diff
@@ -0,0 +1,358 @@
+    return (dataset.shuffle(BUFFER_SIZE).batch(arg.batch_size,
+                                               drop_remainder=True))
+
+
+def make_federated_data(client_data, client_ids, arg):
+    return [
+        preprocess(client_data.create_tf_dataset_for_client(str(x)), arg)
+        for x in client_ids
+    ]
+
+
+def masked_slot_loss(y_true, y_pred):
+    """Defines the loss function to be used while training the model.
+    The loss is defined as the sum of the intent loss and the slot loss
+    per token. Masking is used to not compute the loss on the padding tokens.
+
+    Args:
+    slot_real: The ground truth for the slots.
+    intent_real: The ground through for the intents.
+    slot_pred: The slots predicted by the model.
+    intent_pred: The intents predicted by the model.
+    intent_loss_objective: The objective used to compute the intent loss.
+    slot_loss_objective: The objective used to compute the slot loss.
+
+    Returns:
+    The total loss.
+    """
+    loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
+        from_logits=True, reduction='none')
+    mask = tf.math.logical_not(tf.math.equal(y_true, 0))
+    slot_loss_ = loss_object(y_true, y_pred)
+
+    mask = tf.cast(mask, dtype=slot_loss_.dtype)
+    slot_loss_ *= mask
+    slot_loss = tf.reduce_sum(slot_loss_) / tf.reduce_sum(mask)
+
+    return slot_loss
+
+
+def create_tff_model(arg, input_vocab_size, slot_vocab_size, intent_vocab_size,
+                     input_spec):
+    # TFF uses an `input_spec` so it knows the types and shapes
+    # that your model expects.
+    keras_model = create_keras_model(arg, input_vocab_size, slot_vocab_size,
+                                     intent_vocab_size)
+    return tff.learning.from_keras_model(
+        keras_model,
+        input_spec=input_spec,
+        loss=[
+            masked_slot_loss,
+            tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+        ])
+
+
+def main():
+    """Runs the entire pipelone from loading the data and defining the model
+    to training the model and evaluating the model.
+    """
+
+    arg = parse_arguments()
+
+    in_vocab, slot_vocab, intent_vocab = load_vocab(arg)
+
+    # Loading data
+    train_in_data, train_slot_data, train_intent_data = load_dataset(
+        arg, arg.train_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    # valid_in_data, valid_slot_data, valid_intent_data = load_dataset(
+    #     arg, arg.valid_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    # test_in_data, test_slot_data, test_intent_data = load_dataset(
+    #     arg, arg.test_data_path, in_vocab, slot_vocab, intent_vocab)
+
+    targets = train_slot_data[:, 1:]
+    padding_masks, look_ahead_masks, intent_masks = create_masks(
+        train_in_data, targets)
+
+    num_clients = 10
+    ftrain_data = collections.OrderedDict()
+    instance_per_client = len(train_in_data) // num_clients
+
+    for i in range(num_clients):
```

Comment:
We can submit this as is for now but in a quick follow up change, can you please add some code to persist the split training set per client so that we can do some exploratory analyses on them?

Reply:
I'll just seed the numpy random generator, if that's alright?